### ISSUE TITLE
Templated lambdas crashing the complier

### ISSUE THREAD

**Issue by OgnianM** [llvm/llvm-project#51850]
|  |  |
| --- | --- |
| Bugzilla Link | [52508](https://llvm.org/bz52508) |
| Version | 12.0 |
| OS | Linux |
| CC | @zygoloid |

## Extended Description 
Here's some simplified code that exemplifies the problem.

int test(int x, int y)
{
    auto test_lambda_impl2 = [&]<int f, int z>()
    {
        if constexpr((f + z) < 3)
        {
            cout << f+z << endl;
        }
    };

    auto test_lambda_impl1 = [&]<int f>(int z)
    {
        switch(z)
        {
            case 0:
                test_lambda_impl2.operator()<f, 0>();
                break;
            case 1:
                test_lambda_impl2.operator()<f, 1>();
                break;
            case 2:
                test_lambda_impl2.operator()<f, 2>();
                break;
        }
    };

    switch(x)
    {
        case 0:
            test_lambda_impl1.operator()<0>(y);
            break;
        case 1:
            test_lambda_impl1.operator()<1>(y);
            break;
        case 2:
            test_lambda_impl1.operator()<2>(y);
            break;
    }
}

This just straight up crashes the compiler, if I remove test_lambda_impl2 and just do something in test_lambda_impl1 instead, the code compiles but I get an Invalid instruction exception or a segfault, depending on the code.

Clangd breaks when it has to parse this kind of code as well.

For reference, g++ works as expected.

**Comment by wheatman** [comment]
This seems to have been fixed between 12.0.0 and 13.0.0
https://godbolt.org/z/hsfhbMn49

for reference

code 
```
#include <iostream>
int test(int x, int y)
{
    auto test_lambda_impl2 = [&]<int f, int z>()
    {
        if constexpr((f + z) < 3)
        {
            std::cout << f+z << std::endl;
        }
    };

    auto test_lambda_impl1 = [&]<int f>(int z)
    {
        switch(z)
        {
            case 0:
                test_lambda_impl2.operator()<f, 0>();
                break;
            case 1:
                test_lambda_impl2.operator()<f, 1>();
                break;
            case 2:
                test_lambda_impl2.operator()<f, 2>();
                break;
        }
    };

    switch(x)
    {
        case 0:
            test_lambda_impl1.operator()<0>(y);
            break;
        case 1:
            test_lambda_impl1.operator()<1>(y);
            break;
        case 2:
            test_lambda_impl1.operator()<2>(y);
            break;
    }
}
```

dump

```
<source>:4:33: warning: explicit template parameter list for lambdas is a C++20 extension [-Wc++20-extensions]
    auto test_lambda_impl2 = [&]<int f, int z>()
                                ^
<source>:12:33: warning: explicit template parameter list for lambdas is a C++20 extension [-Wc++20-extensions]
    auto test_lambda_impl1 = [&]<int f>(int z)
                                ^
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-12.0.0/bin/clang++ -gdwarf-4 -g -o /app/output.s -mllvm --x86-asm-syntax=intel -S --gcc-toolchain=/opt/compiler-explorer/gcc-10.3.0 -fcolor-diagnostics -fno-crash-diagnostics -std=c++17 <source>
1.	<source>:31:46: current parser token ')'
2.	<source>:3:1: parsing function body 'test'
3.	<source>:3:1: in compound statement ('{}')
4.	<source>:29:5: in compound statement ('{}')
5.	<source>:12:30: instantiating function definition 'test(int, int)::(anonymous class)::operator()<0>'
6.	<source>:4:30: instantiating class definition ''
 #0 0x000055a1d1dc8bd1 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x318fbd1)
 #1 0x000055a1d1dc6994 llvm::sys::RunSignalHandlers() (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x318d994)
 #2 0x000055a1d1dc6c31 llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x318dc31)
 #3 0x000055a1d1d2f748 CrashRecoverySignalHandler(int) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x30f6748)
 #4 0x00007f0caf184420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #5 0x000055a1d465e4ff clang::Sema::SubstFunctionDeclType(clang::TypeSourceInfo*, clang::MultiLevelTemplateArgumentList const&, clang::SourceLocation, clang::DeclarationName, clang::CXXRecordDecl*, clang::Qualifiers) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a254ff)
 #6 0x000055a1d467165a clang::TemplateDeclInstantiator::SubstFunctionType(clang::FunctionDecl*, llvm::SmallVectorImpl<clang::ParmVarDecl*>&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a3865a)
 #7 0x000055a1d4689bdb clang::TemplateDeclInstantiator::VisitCXXMethodDecl(clang::CXXMethodDecl*, clang::TemplateParameterList*, llvm::Optional<clang::ASTTemplateArgumentListInfo const*>, clang::TemplateDeclInstantiator::RewriteKind) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a50bdb)
 #8 0x000055a1d464f71d clang::Sema::InstantiateClass(clang::SourceLocation, clang::CXXRecordDecl*, clang::CXXRecordDecl*, clang::MultiLevelTemplateArgumentList const&, clang::TemplateSpecializationKind, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a1671d)
 #9 0x000055a1d4682842 clang::TemplateDeclInstantiator::VisitCXXRecordDecl(clang::CXXRecordDecl*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a49842)
#10 0x000055a1d4690dd8 void llvm::function_ref<void ()>::callback_fn<clang::Sema::SubstDecl(clang::Decl*, clang::DeclContext*, clang::MultiLevelTemplateArgumentList const&)::'lambda'()>(long) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a57dd8)
#11 0x000055a1d3e255b5 clang::Sema::runWithSufficientStackSpace(clang::SourceLocation, llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x51ec5b5)
#12 0x000055a1d46711b9 clang::Sema::SubstDecl(clang::Decl*, clang::DeclContext*, clang::MultiLevelTemplateArgumentList const&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a381b9)
#13 0x000055a1d467b4a1 clang::Sema::FindInstantiatedDecl(clang::SourceLocation, clang::NamedDecl*, clang::MultiLevelTemplateArgumentList const&, bool) (.localalias) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a424a1)
#14 0x000055a1d4623d1b (anonymous namespace)::TemplateInstantiator::TransformDecl(clang::SourceLocation, clang::Decl*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x59ead1b)
#15 0x000055a1d465235c clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformUnresolvedMemberExpr(clang::UnresolvedMemberExpr*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a1935c)
#16 0x000055a1d463736b clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformExpr(clang::Expr*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x59fe36b)
#17 0x000055a1d463f834 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformCallExpr(clang::CallExpr*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a06834)
#18 0x000055a1d463798a clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformExpr(clang::Expr*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x59fe98a)
#19 0x000055a1d4665f91 clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformStmt(clang::Stmt*, clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::StmtDiscardKind) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a2cf91)
#20 0x000055a1d466648c clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformCaseStmt(clang::CaseStmt*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a2d48c)
#21 0x000055a1d46676bb clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformCompoundStmt(clang::CompoundStmt*, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a2e6bb)
#22 0x000055a1d466b6fa clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformSwitchStmt(clang::SwitchStmt*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a326fa)
#23 0x000055a1d46676bb clang::TreeTransform<(anonymous namespace)::TemplateInstantiator>::TransformCompoundStmt(clang::CompoundStmt*, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a2e6bb)
#24 0x000055a1d466ba40 clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a32a40)
#25 0x000055a1d4683a50 clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5a4aa50)
#26 0x000055a1d3e255b5 clang::Sema::runWithSufficientStackSpace(clang::SourceLocation, llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x51ec5b5)
#27 0x000055a1d45c07d1 clang::Sema::DeduceReturnType(clang::FunctionDecl*, clang::SourceLocation, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x59877d1)
#28 0x000055a1d417757d clang::Sema::DiagnoseUseOfDecl(clang::NamedDecl*, llvm::ArrayRef<clang::SourceLocation>, clang::ObjCInterfaceDecl const*, bool, bool, clang::ObjCInterfaceDecl*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x553e57d)
#29 0x000055a1d44881ac clang::Sema::BuildCallToMemberFunction(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x584f1ac)
#30 0x000055a1d41c3ca7 clang::Sema::BuildCallExpr(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, bool, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x558aca7)
#31 0x000055a1d41c4c2b clang::Sema::ActOnCallExpr(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x558bc2b)
#32 0x000055a1d3d64d26 clang::Parser::ParsePostfixExpressionSuffix(clang::ActionResult<clang::Expr*, true>) (.localalias) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x512bd26)
#33 0x000055a1d3d5b773 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) (.localalias) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5122773)
#34 0x000055a1d3d5e26e clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, clang::Parser::TypeCastState, bool, bool*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x512526e)
#35 0x000055a1d3d625c9 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x51295c9)
#36 0x000055a1d3d6386d clang::Parser::ParseExpression(clang::Parser::TypeCastState) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x512a86d)
#37 0x000055a1d3dca2d1 clang::Parser::ParseExprStatement(clang::Parser::ParsedStmtContext) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x51912d1)
#38 0x000055a1d3dcaca6 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5191ca6)
#39 0x000055a1d3dcb46c clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519246c)
#40 0x000055a1d3dcb702 clang::Parser::ParseStatement(clang::SourceLocation*, clang::Parser::ParsedStmtContext) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5192702)
#41 0x000055a1d3dccf76 clang::Parser::ParseCaseStatement(clang::Parser::ParsedStmtContext, bool, clang::ActionResult<clang::Expr*, true>) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5193f76)
#42 0x000055a1d3dcb17b clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519217b)
#43 0x000055a1d3dcb46c clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519246c)
#44 0x000055a1d3dd4841 clang::Parser::ParseCompoundStatementBody(bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519b841)
#45 0x000055a1d3dd52bf clang::Parser::ParseCompoundStatement(bool, unsigned int) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519c2bf)
#46 0x000055a1d3dcaaea clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5191aea)
#47 0x000055a1d3dcb46c clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519246c)
#48 0x000055a1d3dcb702 clang::Parser::ParseStatement(clang::SourceLocation*, clang::Parser::ParsedStmtContext) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x5192702)
#49 0x000055a1d3dd05bd clang::Parser::ParseSwitchStatement(clang::SourceLocation*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x51975bd)
#50 0x000055a1d3dcb0db clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x51920db)
#51 0x000055a1d3dcb46c clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519246c)
#52 0x000055a1d3dd4841 clang::Parser::ParseCompoundStatementBody(bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519b841)
#53 0x000055a1d3dd8372 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x519f372)
#54 0x000055a1d3d07a6a clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x50cea6a)
#55 0x000055a1d3d3585f clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x50fc85f)
#56 0x000055a1d3d01f22 clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x50c8f22)
#57 0x000055a1d3d02555 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (.part.0) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x50c9555)
#58 0x000055a1d3d0bd70 clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) (.localalias) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x50d2d70)
#59 0x000055a1d3d0c0b5 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x50d30b5)
#60 0x000055a1d3cfb2e9 clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x50c22e9)
#61 0x000055a1d2db0626 clang::CodeGenAction::ExecuteAction() (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x4177626)
#62 0x000055a1d270aab1 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x3ad1ab1)
#63 0x000055a1d26a7ebb clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x3a6eebb)
#64 0x000055a1d27dc940 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x3ba3940)
#65 0x000055a1cfbbf5f1 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0xf865f1)
#66 0x000055a1cfbbcbbb ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0xf83bbb)
#67 0x000055a1d254e299 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const::'lambda'()>(long) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x3915299)
#68 0x000055a1d1d2f847 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x30f6847)
#69 0x000055a1d254ed96 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const (.part.0) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x3915d96)
#70 0x000055a1d252322c clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x38ea22c)
#71 0x000055a1d2523c49 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x38eac49)
#72 0x000055a1d252f381 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0x38f6381)
#73 0x000055a1cfaff41a main (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0xec641a)
#74 0x00007f0caec32083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#75 0x000055a1cfbbc74e _start (/opt/compiler-explorer/clang-12.0.0/bin/clang+++0xf8374e)
clang-12: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Compiler returned: 139
```

**Comment by Endilll** [comment]
Slightly modified to silence diagnostics: https://godbolt.org/z/hKjsszTTn
